packages/core/strapi/lib/services/entity-service/types/result.d.ts

```diff
@@ -36,3 +75,133 @@ type GetValue<TAttribute extends Attribute.Attribute> = Utils.Expression.If<
     Attribute.GetValue<TAttribute>
   >
 >;
+
+/**
+ * attributes values => <uid, fields>
+ */
+
+// transform fields (string | string[]) & populate (object, string, string[]) -> string union (keys)
+
+type As<T, P> = T extends P ? T : never;
+
+type ExtractFields<
+  TSchemaUID extends Common.UID.Schema,
+  TFields extends EntityService.Params.Fields.Any<TSchemaUID>
+> = Utils.Expression.MatchFirst<
+  [
+    // string
+    [
+      Utils.Expression.Extends<TFields, EntityService.Params.Fields.StringNotation<TSchemaUID>>,
+      ParseStringFields<
+        TSchemaUID,
+        As<TFields, EntityService.Params.Fields.StringNotation<TSchemaUID>>
+      >
+    ],
+    // string array
+    [
+      Utils.Expression.Extends<TFields, EntityService.Params.Fields.ArrayNotation<TSchemaUID>>,
+      ParseStringFields<
+        TSchemaUID,
+        As<
+          Utils.Array.Values<As<TFields, EntityService.Params.Fields.ArrayNotation<TSchemaUID>>>,
+          EntityService.Params.Fields.StringNotation<TSchemaUID>
+        >
+      >
+    ]
+  ]
+>;
+
+type ParseStringFields<
+  TSchemaUID extends Common.UID.Schema,
+  TFields extends EntityService.Params.Fields.StringNotation<TSchemaUID>
+> = Utils.Expression.MatchFirst<
+  [
+    [Utils.Expression.IsNever<TFields>, Attribute.GetNonPopulatableKeys<TSchemaUID>],
+    [
+      Utils.Expression.Extends<EntityService.Params.Fields.WildcardNotation, TFields>,
+      Attribute.GetNonPopulatableKeys<TSchemaUID>
+    ],
+    [
+      Utils.Expression.Extends<TFields, EntityService.Params.Fields.SingleAttribute<TSchemaUID>>,
+      TFields
+    ],
+    [
+      // what aobut * in comma separated list
```

Comment:
Where are we regarding this?

```suggestion
      // what about * in comma separated list
```